### ISSUE TITLE
Pyramid level calculation for multiscale is incorrect

### ISSUE THREAD

**Issue by michael-sutherland** [napari/napari#4781]
## 🐛 Bug

When using pyramids in Napari, the level chosen appears to be incorrect. This can cause the application to slow down or crash in certain instances.  Specifically, a lower resolution level should often be chosen than what is actual picked.  The main issue appears to be that the algorithm assumes the layer image fills the entire field of view when picking the pyramid level.  I think this may also be a contributing factor into why the multiscale images are slower than they should be in napari.

## To Reproduce

1.  Create pyramid view (I'm using a 2D pyramid) with multiple levels (say 1kx1k with 4 levels)
2.  Zoom out 
3.  Observe the level chosen (it is in the toolbar).  For an image smaller than the display, it will never go below 0 even if you zoom way out.

Issue I'm running into currently that crashes napari:
I have a large mosaic image that is not square.  The image has dimensions of [3072, 70656].  When I zoom out, napari picks the zoom level to be 1 (scale is [2,2]).  This is because the y-axis is being compared to the whole view window, even though it is only taking up a small area of the window. The x-axis at level 1 is now greater than the max texture size and napari crashes.

## Expected behavior

Napari should NOT use the display window size by itself when calculating the level, but should calculate the extent of the layer being displayed and use that instead.  Then it can pick the level of the multiscale that most closely matches the what needs to be displayed.  To put in other words, if we zoom out so that the 10kx10k image now only takes up 100px x 100px on the display, we should pick the level that is closest to the 100x100 pixels instead of using the size of the display.

napari/layers/base/base.py Layer._update_draw function:
The parameter shape_threshold needs to be cropped to match the extent of the layer being displayed before being passed to compute_multiscale_level.  It looks like the code crops the size of the layer to only what is being displayed, but then doesn't crop the shape_threshold to the layer as well.

## Environment

napari: 0.4.16
Platform: Linux-4.18.0-372.9.1.el8.x86_64-x86_64-with-glibc2.28
Python: 3.9.0 (default, Nov 15 2020, 14:28:56) [GCC 7.3.0]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.22.3
SciPy: 1.8.0
Dask: 2022.05.0
VisPy: 0.10.0

OpenGL:
- GL version: 4.6.0 NVIDIA 515.48.07
- MAX_TEXTURE_SIZE: 32768

Screens:
- screen 1: resolution 3840x2160, scale 1.0

Plugins:
- console: 0.0.4
- napari-svg: 0.1.6
- scikit-image: 0.4.16

 - Any other relevant information:

**Comment by jni** [comment]
Good point @michael-sutherland and I believe I've encountered it before myself. Agree this needs fixing.

**Comment by andy-sweet** [comment]
Thanks for the write-up here. I'm going to assign myself to this issue because there's some strong overlap with the async slicing redesign I'm working on - we need the non-experimental multi-scale data lookup to work as intended for some of that work to be useful.

I'll start by writing up some test cases based on the great description above.

Thanks @michael-sutherland!

**Comment by michael-sutherland** [comment]
Depending on timelines, it might be nice to patch the potential crashing.  I threw in an ugly check to make sure I'm not picking a level that is displaying more pixels than the max texture size.  I ended up hard coding the texture size for my computer to make it easy for me.  Here's my snippet from layers/utils/layer_utils.py:

```python
def compute_multiscale_level(
    requested_shape, shape_threshold, downsample_factors
):
    """Computed desired level of the multiscale given requested field of view.

    The level of the multiscale should be the lowest resolution such that
    the requested shape is above the shape threshold. By passing a shape
    threshold corresponding to the shape of the canvas on the screen this
    ensures that we have at least one data pixel per screen pixel, but no
    more than we need.

    Parameters
    ----------
    requested_shape : tuple
        Requested shape of field of view in data coordinates
    shape_threshold : tuple
        Maximum size of a displayed tile in pixels.
    downsample_factors : list of tuple
        Downsampling factors for each level of the multiscale. Must be increasing
        for each level of the multiscale.

    Returns
    -------
    level : int
        Level of the multiscale to be viewing.
    """
    # Scale shape by downsample factors
    scaled_shape = requested_shape / downsample_factors

    # HACK: find min level that doesn't exceed size limit
    min_level = 0
    for level in range(scaled_shape.shape[0]):
        if np.any(scaled_shape[level] >= 32768):
            min_level = level + 1

    # Find the highest resolution level allowed
    locations = np.argwhere(np.all(scaled_shape > shape_threshold, axis=1))
    if len(locations) > 0:
        level = locations[-1][0]
    else:
        level = 0

    level = max(min_level, level)   # HACK
    return level
```

**Comment by andy-sweet** [comment]
I'm actually surprised by the crashing behavior, as [napari has some existing logic to downsample any image that exceeds that the max texture size](https://github.com/napari/napari/blob/67c50a4bec99d6a0f381276195ccc1ee1210b876/napari/_vispy/layers/image.py#L129), though it is less efficient than your approach. Maybe there's at least one case when that doesn't happen? I plan to add to add at least one test case to cover that.

Thanks for the suggested workaround. Ideally, I'd like to have a more complete fix here with some test coverage. I think I almost have a suitable test case to cover the issues here, though it seems like I have to do some special setup to replicate errors in napari-proper.

```python
@pytest.mark.parametrize(
    ('camera_zoom', 'expected_data_level'), ((1, 0), (0.5, 1), (0.25, 2))
)
def test_on_draw_with_multiscale_image_camera_zoom(
    make_napari_viewer, camera_zoom, expected_data_level
):
    viewer = make_napari_viewer()
    qt_viewer = viewer.window._qt_viewer
    shapes = [(1600, 1600), (800, 800), (400, 400)]
    data = [np.empty(s) for s in shapes]
    layer = viewer.add_image(data, multiscale=True)
    qt_viewer.canvas.size = (1000, 1000)
    # This does not get triggered automatically in the test.
    qt_viewer.on_resize(None)
    viewer.reset_view()
    viewer.camera.zoom = camera_zoom

    qt_viewer.on_draw(None)

    assert layer.data_level == expected_data_level
```

If I don't have a PR that fixes this test by the end of the week, then I'll definitely consider bringing in your suggested workaround, which could make it into the next patch version of napari.

**Comment by andy-sweet** [comment]
@michael-sutherland : took me a little longer than I said, but I think I have a fix for this in #4842 

Feel free to comment on or review that PR and/or try that branch out to see if it fixes the issue for you.
